flink-kubernetes-operator/src/main/java/org/apache/flink/kubernetes/operator/reconciler/deployment/ApplicationReconciler.java

```diff
@@ -111,12 +114,34 @@ public void reconcile(FlinkDeployment flinkApp, Context context, Configuration e
             }
             IngressUtils.updateIngressRules(flinkApp, effectiveConfig, kubernetesClient);
             ReconciliationUtils.updateForSpecReconciliationSuccess(flinkApp, stateAfterReconcile);
+        } else if (ReconciliationUtils.shouldRollBack(flinkApp)) {
+            rollbackApplication(flinkApp);
         } else if (SavepointUtils.shouldTriggerSavepoint(flinkApp) && isJobRunning(flinkApp)) {
             triggerSavepoint(flinkApp, effectiveConfig);
             ReconciliationUtils.updateSavepointReconciliationSuccess(flinkApp);
         }
     }
 
+    private void rollbackApplication(FlinkDeployment flinkApp) throws Exception {
+        LOG.warn("Rolling back deployment to last stable spec");
+        Configuration rollbackConfig =
+                FlinkUtils.getEffectiveConfig(
+                        flinkApp.getMetadata(),
+                        flinkApp.getStatus().getReconciliationStatus().deserializeLastStableSpec(),
+                        defaultConfig.getFlinkConfig());
+        suspendJob(flinkApp, UpgradeMode.LAST_STATE, rollbackConfig);
+        deployFlinkJob(
+                flinkApp,
+                rollbackConfig,
+                Optional.ofNullable(
+                                flinkApp.getStatus()
+                                        .getJobStatus()
+                                        .getSavepointInfo()
+                                        .getLastSavepoint())
+                        .map(Savepoint::getLocation));
+        flinkApp.getStatus().getReconciliationStatus().setRolledBack(true);
```

Comment:
Do we have to update the `lastReconciledSpec` to `lastStableSpec` after rollback?

Reply:
Oh, I get it now. It will leave the `lastReconciledSpec = currentSpec`, but mark the `lastReconciledSpec` have be rollbacked

Reply:
Yes, this way it is clear from the status what happened.